### ISSUE TITLE
Allow AsyncLazy to be reset

### ISSUE THREAD

**Issue by CharlesWillis3** [microsoft/vs-threading#564]
Imagine using AsyncLazy to defer getting a service instance. For your scenario, a one time initialization is usually all that is required. You find that occasionally some external state change requires that the service will be need to fetched again. You don't want to actually incur the cost of fetching the service at the point the external state change is noticed. You want to ensure that the next time the service is called, it's been fetched again.

If you have a field that holds the AsyncLazy reference, you can accomplish this by constructing a new AsyncLazy and pointing the field to it. But then access to the field may need to be synchronized.

What if AsyncLazy could be reset, so that a future call to GetValueAsync() would run the factory again?

**Comment by sharwell** [comment]
State updates of this form tend to be extremely complicated. I do not believe a 'reset' would provide a clear improvement for any given scenario, and introduces other serious pitfalls (what happens if an instance gets reset that was expected to be a singleton?).

"Changing the service" means all outstanding operations using the old service need to cleanly roll forward to the new service. Does this happen in parallel with the initialization of the new service, or does the new instance need to wait for all uses of the old instance to release it?

**Comment by AArnott** [comment]
I don't think this feature should try to absolve the owner from all responsibility regarding side effects of the factory. But I have occasionally hit myself the case where I want to reinitialize the AsyncLazy, but to do that I have to basically copy the definition of the value factory from my own class constructor to the method that wants to reinitialize. At this point, I have to move the body of the value factory into its own method, change the ctor to use that method and then use that method elsewhere. Oh, and I have to make sure I pass JTF in at both places (or not, depending). It would all be so much simpler (and semantically more clear) if I could just call .Reset() on it.
Any prior value or in-progress execution of the value factory would continue as-is. It would be exactly the same behaviorally as if I just reassigned the AsyncLazy field.

**Comment by AArnott** [comment]
Part of AsyncLazy's internal synchronization relies on clearing of the value factory field, so resetting state would require that we hang onto that value factory long-term, which would have at least a small memory cost as well as requiring *very* careful work to revise how we do our internal synchronization in the class as we find a substitute means to do it.
Given that and only one request for this feature, I'm going to Won't Fix this. We can reconsider if there's business justification or this becomes a more popular request.
